src/Domain/Services/PayForOrder.php

```diff
@@ -0,0 +1,51 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+/**
+ * Service class implementing authentication checks for pay-for-order checkout block.
+ */
+class PayForOrder {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->init();
+	}
+
+	/**
+	 * Hook into WP.
+	 */
+	public function init() {
+		add_filter( 'the_content', array( $this, 'check_authentication' ) );
+	}
+
+	/**
+	 * Check authentication for pay for order
+	 *
+	 * @param string $content Content.
+	 */
+	public function check_authentication( $content ) {
+		global $wp;
+
+		$order_id = ! empty( $wp->query_vars['order-pay'] ) ? $wp->query_vars['order-pay'] : null;
+
+		if ( isset( $_GET['pay_for_order'], $_GET['key'] ) && $order_id ) { // phpcs:ignore WordPress.Security.NonceVerification
```

Comment:
This is looking to see if `pay_for_order` is set, but its not validating it against the actual order key.

Reply:
Since we already validate the order key in [this PR](https://github.com/woocommerce/woocommerce-blocks/pull/8960/files) so we don't need to validate it again, right?